lua/bufferline/events.lua

```diff
@@ -0,0 +1,248 @@
+    group = augroup_render,
+    pattern = 'buflisted',
+  })
+
+  create_autocmd('SessionLoadPost', {
+    -- TODO: I'm not sure if this whole thing can just be thrown in a `schedule_wrap`
+    --       in order to remove the inner `defer_fn` and `schedule` calls
+    callback = function()
+      if state.loading_session then
+        return
+      end
+
+      local restore_cmd = vim.g.Bufferline__session_restore
+      if restore_cmd then command(restore_cmd) end
+
+      vim.defer_fn(function() state.loading_session = false end, 100)
+      schedule(function() render.update(true) end)
+    end,
+    group = augroup_render,
+  })
+
+  create_autocmd('TermOpen', {
+    callback = function() defer_fn(function() render.update(true) end, 500) end,
+    group = augroup_render,
+  })
+
+  create_autocmd('User', {
+    callback = function()
+      -- We're allowed to use relative paths for buffers iff there are no tabpages
+      -- or windows with a local directory (:tcd and :lcd)
+      local use_relative_file_paths = true
+
+      -- PERF: I didn't import `haslocaldir` since it is only used here (just before calling `:mksession` and exiting vim)
+      local haslocaldir = vim.fn.haslocaldir --- @type function
+
+      for tabnr, tabpage in ipairs(list_tabpages()) do
+        if not use_relative_file_paths or haslocaldir(-1, tabnr) == 1 then
+          use_relative_file_paths = false
+          break
+        end
+
+        for _, win in ipairs(tabpage_list_wins(tabpage)) do
+          if haslocaldir(win, tabnr) == 1 then
+            use_relative_file_paths = false
+            break
+          end
+        end
+      end
+
+      local buffers = {}
+      for _, bufnr in ipairs(state.buffers) do
+        local name = buf_get_name(bufnr)
+        if use_relative_file_paths then
+          name = utils.relative(name)
+        end
+
+        -- escape quotes
+        table_insert(buffers, {name = name, pinned =  state.get_buffer_data(bufnr).pinned})
+      end
+
+      vim.g.Bufferline__session_restore = "lua require'bufferline.state'.restore_buffers " ..
+        vim.inspect(buffers, {newline = ' ', indent = ''})
+    end,
+    group = augroup_misc,
+    pattern = 'SessionSavePre',
+  })
+
+  create_autocmd('WinClosed', {
+    callback = schedule_wrap(render.update),
+    group = augroup_render,
+  })
+
+  vim.schedule(function()
+    dictwatcherdel()
+    vim.cmd [[
+      " Must be global -_-
+      function! BufferlineOnOptionChanged(dict) abort
+        call luaeval("require'bufferline.events'.on_option_changed(_A)", a:dict)
+      endfunction
+
+      " Must be global -_-
+      function! BufferlineDictChanged(dict, key, changes) abort
+        silent! call dictwatcherdel(a:changes.old, '*', 'BufferlineOnOptionChanged')
+        call dictwatcheradd(a:changes.new, '*', 'BufferlineOnOptionChanged')
+        call BufferlineOnOptionChanged(a:changes.new)
+      endfunction
+
+      call dictwatcheradd(g:, 'bufferline', 'BufferlineDictChanged')
```

Comment:
Here's an interesting change. 

I found out while testing that, intermittently, the `dictwatcher` for `g:bufferline` would stop working. I found out that this was because `g:bufferline` was being _replaced_, and when this happened, the `dictwatcher` was looking for changes on the old dict. Example:

```vim
let g:bufferline = {}
call dictwatcheradd(g:bufferline, '*', 'BufferlineOnOptionChanged')
let g:bufferline.foo = v:true " dictwatcher responds
let g:bufferline = {'bar': v:false} " dictwatcher does nothing
```

This can be fixed by listening to `g:` itself with the `'bufferline'` pattern, and call `dictwatcheradd` on the new `g:bufferline`.

Now, the `on_option_changed` callback will be run much more consistently than before.